File: cpp/567_permutation_in_string.cpp
#include <array>
#include <string>

using namespace std;

class Solution {
 public:
  bool checkInclusion(string s1, string s2) {
    int len1 = s1.size(), len2 = s2.size();
    if (len1 > len2) {
      return false;
    }

    array<int, 26> letter_freq{};
    for (const auto c : s1) {
      letter_freq[c - 'a']++;
    }

    int diff = len1;
    for (int end = 0; end < s2.size(); end++) {
      int start = end - len1;
      if (start >= 0) {
        char c = s2[start];
        // If a letter doesn't appear in s1 or appear in s2
        // more than in s1, its frequency is less than 0
        if (letter_freq[c - 'a'] >= 0) {
          diff++;
        }
        letter_freq[c - 'a']++;
      }

      char c = s2[end];
      // If this letter appears in s1, decrement diff
      if (letter_freq[c - 'a'] > 0) {
        diff--;
      }
      letter_freq[c - 'a']--;
      if (diff == 0) {
        return true;
      }
    }

    return false;
  }
};

File: rust/567_permutation_in_string.rs
impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let (s1, s2) = (s1.as_bytes(), s2.as_bytes());
        let (len1, len2) = (s1.len(), s2.len());
        if len1 > len2 {
            return false;
        }

        let mut letter_freq = [0; 26];
        for c in s1 {
            letter_freq[(c - b'a') as usize] += 1
        }

        let mut diff = len1;
        for end in 0..len2 {
            if end >= len1 {
                let c = s2[end - len1];
                if letter_freq[(c - b'a') as usize] >= 0 {
                    diff += 1;
                }
                letter_freq[(c - b'a') as usize] += 1;
            }

            let c = s2[end];
            if letter_freq[(c - b'a') as usize] > 0 {
                diff -= 1;
            }
            letter_freq[(c - b'a') as usize] -= 1;
            if diff == 0 {
                return true;
            }
        }

        false
    }
}
